File: copernicoaches/atrium/atrium/state.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from .filesystem import FilesystemAdapter
from .models import Entry


@dataclass(slots=True)
class AtriumState:
    current_dir: Path
    parent_dir: Path | None
    current_entries: tuple[Entry, ...]
    selected: Entry | None
    preview_target: Entry | None
    current_error: str | None = None


def build_initial_state(
    start_path: Path | None = None,
    *,
    filesystem: FilesystemAdapter | None = None,
) -> AtriumState:
    adapter = filesystem or FilesystemAdapter()
    current_dir = (start_path or Path.cwd()).resolve()
    listing = adapter.list_directory(current_dir)
    selected = listing.entries[0] if listing.entries else None

    return AtriumState(
        current_dir=current_dir,
        parent_dir=current_dir.parent if current_dir.parent != current_dir else None,
        current_entries=listing.entries,
        selected=selected,
        preview_target=selected,
        current_error=listing.error,
    )

File: copernicoaches/atrium/src/state.rs
use std::env;
use std::path::PathBuf;

use crate::filesystem::Filesystem;
use crate::models::Entry;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AppState {
    pub current_dir: PathBuf,
    pub parent_dir: Option<PathBuf>,
    pub current_entries: Vec<Entry>,
    pub selected: Option<Entry>,
    pub preview_target: Option<Entry>,
    pub current_error: Option<String>,
}

pub fn build_initial_state(start_path: Option<PathBuf>, filesystem: &impl Filesystem) -> AppState {
    let start_path = start_path.unwrap_or_else(|| env::current_dir().expect("current directory"));
    let listing = filesystem.list_directory(&start_path);
    let current_dir = listing.path;
    let selected = listing.entries.first().cloned();

    AppState {
        parent_dir: current_dir.parent().map(|path| path.to_path_buf()),
        current_dir,
        current_entries: listing.entries,
        selected: selected.clone(),
        preview_target: selected,
        current_error: listing.error,
    }
}
